metadata/src/main/java/org/apache/kafka/metadata/storage/Formatter.java

```diff
@@ -0,0 +1,506 @@
+        ensemble.verify(Optional.of(clusterId),
+                OptionalInt.of(nodeId),
+                EnumSet.noneOf(MetaPropertiesEnsemble.VerificationFlag.class));
+        MetaPropertiesEnsemble.Copier copier = new MetaPropertiesEnsemble.Copier(ensemble);
+        if (!(ignoreFormatted || copier.logDirProps().isEmpty())) {
+            String firstLogDir = copier.logDirProps().keySet().iterator().next();
+            throw new FormatterException("Log directory " + firstLogDir + " is already formatted. " +
+                "Use --ignore-formatted to ignore this directory and format the others.");
+        }
+        if (!copier.errorLogDirs().isEmpty()) {
+            copier.errorLogDirs().forEach(errorLogDir ->
+                printStream.println("I/O error trying to read log directory " + errorLogDir + ". Ignoring..."));
+            if (ensemble.emptyLogDirs().isEmpty() && copier.logDirProps().isEmpty()) {
+                throw new FormatterException("No available log directories to format.");
+            }
+        }
+        if (ensemble.emptyLogDirs().isEmpty()) {
+            printStream.println("All of the log directories are already formatted.");
+        } else {
+            Map<String, DirectoryType> directoryTypes = new HashMap<>();
+            for (String emptyLogDir : ensemble.emptyLogDirs()) {
+                DirectoryType directoryType = DirectoryType.calculate(emptyLogDir,
+                    metadataLogDirectory,
+                    nodeId,
+                    initialControllers);
+                directoryTypes.put(emptyLogDir, directoryType);
+                Uuid directoryId;
+                if (directoryType == DirectoryType.DYNAMIC_METADATA_VOTER_DIRECTORY) {
+                    directoryId = initialControllers.get().voters().get(nodeId).directoryId();
+                } else {
+                    directoryId = copier.generateValidDirectoryId();
+                }
+                copier.setLogDirProps(emptyLogDir, new MetaProperties.Builder(metaProperties).
+                    setDirectoryId(directoryId).
+                    build());
+            }
+            copier.setPreWriteHandler((writeLogDir, __, ____) -> {
+                printStream.printf("Formatting %s %s with %s %s.%n",
+                    directoryTypes.get(writeLogDir).description(), writeLogDir,
+                    MetadataVersion.FEATURE_NAME, releaseVersion);
+                Files.createDirectories(Paths.get(writeLogDir));
+                BootstrapDirectory bootstrapDirectory = new BootstrapDirectory(writeLogDir, Optional.empty());
+                bootstrapDirectory.writeBinaryFile(bootstrapMetadata);
+                if (directoryTypes.get(writeLogDir).isDynamicMetadataDirectory()) {
+                    writeDynamicQuorumSnapshot(writeLogDir,
+                        initialControllers.get(),
+                        featureLevels.get(KRaftVersion.FEATURE_NAME),
+                        controllerListenerName);
```

Comment:
Okay. So the kraft.version = 0 if and only if there is are no `DynamicVoter`/`VoterSet` and the tool don't write a `0-0.checkpoint` file. The kraft.version = 1 if and only if there are `DynamicVoter`/`VoterSet` and the tool write a `0-0.checkpoint` file.

Is that correct?

Reply:
yes, that's correct